### PR TITLE
[WIP] Port packet handling updates

### PR DESCRIPTION
**Pull request by ariel-**
6f65d48a2149d4b5145f664bbf1aae20c1ea0b27 and others.

### CODE REVIEW

**Comment by Shauren** [comment]
mmmm, not sure about this in 3.3.5 brach, its a lot of work that does not need to be done (unlike 6.x where packet structures change)

**Comment by DDuarte** [comment]
I think this is great on the 3.3.5 branch: consistency in the branches, less code duplication, single place to write/read packets, etc

**Comment by Naios** [comment]
I thought about replacing our hardcoded Optional class with boost::optional (https://github.com/Naios/TrinityCore/commits/optional) which requires some changes on the packets, maybe we finish it first before porting everything to 3.3.5. Currently i don't have that much time to finish the branch.

**Comment by Shauren** [comment]
@Naios that is honestly way overkill - why replacing a incredibly simple 20 line structure with this monster spanning across MULTIPLE headers http://www.boost.org/doc/libs/1_58_0/boost/optional/optional.hpp

**Comment by Naios** [comment]
To make it compatible with the upcoming c++14 optional, also it is empty optimized and supports r-value references.

**Comment by ccrs** [comment]
OMG this guy :O 

Have my babies pls
image

**Comment by Naios** [comment]
As already mentioned i replaced the Optional by boost::optional in 800d5d893964a8226.
Could you include the changes made there please?

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -5053,20 +4846,27 @@ void Player::DeleteOldCharacters(uint32 keepDays)
...
 {
-    WorldPacket data;
+    WorldPackets::ServerPacket* packet = nullptr;
     switch (pType)
     {
-        case MOVE_ROOT:       data.Initialize(SMSG_FORCE_MOVE_ROOT,   GetPackGUID().size()+4); break;
-        case MOVE_UNROOT:     data.Initialize(SMSG_FORCE_MOVE_UNROOT, GetPackGUID().size()+4); break;
-        case MOVE_WATER_WALK: data.Initialize(SMSG_MOVE_WATER_WALK,   GetPackGUID().size()+4); break;
-        case MOVE_LAND_WALK:  data.Initialize(SMSG_MOVE_LAND_WALK,    GetPackGUID().size()+4); break;
+        case MOVE_ROOT:       packet = new WorldPackets::Movement::MoveForceRoot; break;
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: bah. memory leak.

**Comment by Shauren** [comment]
@Naios 3.3.5 does not have the concept of Optional in packets

**Comment by Naios** [comment]
But optionals are used in this Pr.

**Comment by Shauren** [comment]
well, they shouldnt be

**Comment by ariel-** [comment]
@Shauren could you elaborate more on this?

**Comment by jackpoz** [comment]
out of curiousity, how much of the modified code has been debugged to verify everything works correctly ?

**Comment by ariel-** [comment]
I took a break from this one, it's true that its too much work, but when I left, it was working as far as tests could go (logged two accounts and tested movement).

If you'd like, I can solve the now existing conflicts and give it a go (since even on 6.x  branch some packets are still handled the old way)

**Comment by jackpoz** [comment]
I'm worried about breaking any current working functionality in 335, especially with 108 files changed

**Comment by jackpoz** [comment]
https://github.com/TrinityCore/TrinityCore/pull/14924#issuecomment-113886099 makes me wonder if there has been any similar change to 6.x packet handling that changed code behavior

File: src/server/game/Entities/Player/Player.cpp (hunk 1 above)
Comment (jackpoz) [inline src/server/game/Entities/Player/Player.cpp]: @ariel- can you take a look at this ?

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -5053,20 +4846,28 @@ void Player::DeleteOldCharacters(uint32 keepDays)
...
+    switch (pType)
+    {
+        case MOVE_ROOT: reinterpret_cast<WorldPackets::Movement::MoveForceRoot*>(packet)->Guid = GetPackGUID(); break;
+        case MOVE_UNROOT: reinterpret_cast<WorldPackets::Movement::MoveForceUnroot*>(packet)->Guid = GetPackGUID(); break;
+        case MOVE_WATER_WALK: reinterpret_cast<WorldPackets::Movement::MoveForceWaterWalk*>(packet)->Guid = GetPackGUID(); break;
+        case MOVE_LAND_WALK: reinterpret_cast<WorldPackets::Movement::MoveForceLandWalk*>(packet)->Guid = GetPackGUID(); break;
+    }
+    
+    GetSession()->SendPacket(packet->Write());
+    delete packet;
```
Comment (ariel-) [inline src/server/game/Entities/Player/Player.cpp]: @jackpoz, the delete was missing back then, here it is added (packet is not null at this point)

File: src/server/game/Entities/Player/Player.cpp (hunk 1 above)
Comment (jackpoz) [inline src/server/game/Entities/Player/Player.cpp]: fixed at https://github.com/TrinityCore/TrinityCore/pull/14562#discussion_r34221487

**Comment by jackpoz** [comment]
this PR has so many changes and affects so many packets that is simply impossible to review.
thank you for the work, it will be most likely useful when a developer will choose to port 6.x packet handling to 335 too

**Comment by ariel-** [comment]
Superceded by #17777

**Pull request closed without merging**